server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -608,7 +608,11 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
     public void relocated(final Consumer<ReplicationTracker.PrimaryContext> consumer)
                                             throws IllegalIndexShardStateException, InterruptedException {
         assert shardRouting.primary() : "only primaries can be marked as relocated: " + shardRouting;
+        refreshListeners.disallowAdd();
```

Comment:
could this be concurrently called? What should the semantics then be? I think we'll need a `refreshForcers` counter in `RefreshListeners`, a simple boolean might not suffice

Reply:
You're right I think, moved to counter :)